### PR TITLE
[js-api] Define multiple return value semantics

### PR DESCRIPTION
**Pull request by littledan**
- When a JS function is imported by Wasm, with a multiple return
  value type, its return value is iterated over to get the multiple
  WebAssembly values.
- When a WebAssembly function is exported to JS, multiple return
  values become an Array of those values.

These semantics were approved by a poll at the 2018 WebAssembly CG
TPAC meeting.

The wording here is based on JS operations rather than WebIDL;
this is because the operations ToJSValue/ToWebAssemblyValue are
based on the JS type Number rather than IDL types like double.
However, the semantics are in parallel with what it means to
convert from and to a WebIDL sequence.

### CODE REVIEW

File: document/js-api/index.bs
Comment (rossberg) [inline document/js-api/index.bs]: s/imports/wasmValues/ ?

File: document/js-api/index.bs
Comment (rossberg) [inline document/js-api/index.bs]: wasmValues? However, you cannot return a list to Wasm. You'll need to convert it to a value sequence suitable for the core spec.

**Review by rossberg — COMMENTED** [review]
Thanks!

File: document/js-api/index.bs
Comment (littledan) [reply document/js-api/index.bs]: Done

File: document/js-api/index.bs
Comment (littledan) [reply document/js-api/index.bs]: Switched to wasmValues.

Do you have a suggestion for the wording for exposing a sequence to the core spec? Note that this spec already uses infra lists and Wasm sequences interchangably, as in the _args_ list in the "call an Exported Function" algorithm.

Comment (rossberg) [reply document/js-api/index.bs]: I see, that's unfortunate. Perhaps simply introduce a meta operator ToWasmSeq that turns a list into a sequence somewhere? They cannot be interchangeable because the notion of mutation does not make sense for a syntactic sequence.

Comment (littledan) [reply document/js-api/index.bs]: That makes sense. I'll do this in a separate patch against the main spec repo.

File: document/js-api/index.bs
Comment (xtuc) [inline document/js-api/index.bs]: nit: could we rename it to `retArity`?

File: document/js-api/index.bs
Comment (littledan) [reply document/js-api/index.bs]: Landed the other patch, and used the same wording to fix this case.

File: document/js-api/index.bs
Comment (littledan) [reply document/js-api/index.bs]: Feel free to propose that in a different PR (maybe against the main branch). outArity isn't a new name here.

**Comment by xtuc** [comment]
We could already add an entry into the implementation limits. The maxium number of params is 1000, the maxium number of returns could be the same.

On a side note, the event https://github.com/WebAssembly/multi-value/pull/8#event-2209870726 was generated by me because I delete my comment where @littledan's comment was attached to.

**Comment by littledan** [comment]
Adding an implementation limit sounds like a good idea. @xtuc, Would you be interested in making a PR against this branch that adds that?

File: document/js-api/index.bs
Comment (rossberg) [inline document/js-api/index.bs]: Nit: To be pedantic, this wording is ambiguous, since it could either describe a join or a cross product.

Also, maybe add an assert that both have the same arity?

**Review by rossberg — APPROVED** [review]

**Review by xtuc — APPROVED** [review]
LGTM!

Once this is merged, let's maybe bring the proposal to the WebAssembly GC's attention again.

File: document/js-api/index.bs
Comment (littledan) [reply document/js-api/index.bs]: Good point. I'm not sure what wording is the best; what do you think of this?

Comment (rossberg) [reply document/js-api/index.bs]: LGTM

**Comment by littledan** [comment]
Well, I don't have write access. Feel free to merge this PR, @rossberg .

@Ms2ger Further reviews (and tests) welcome.

**Comment by rossberg** [comment]
Thanks!

**Comment by Ms2ger** [comment]
Some tests: https://github.com/web-platform-tests/wpt/pull/17314

**Pull request merged**
